src/hotspot/share/runtime/handshake.cpp

```diff
@@ -30,37 +30,44 @@
 #include "runtime/handshake.hpp"
 #include "runtime/interfaceSupport.inline.hpp"
 #include "runtime/osThread.hpp"
-#include "runtime/semaphore.inline.hpp"
 #include "runtime/task.hpp"
 #include "runtime/thread.hpp"
 #include "runtime/vmThread.hpp"
 #include "utilities/formatBuffer.hpp"
+#include "utilities/filterQueue.inline.hpp"
 #include "utilities/preserveException.hpp"
 
-
-class HandshakeOperation: public StackObj {
-  HandshakeClosure* _handshake_cl;
-  int32_t _pending_threads;
-  bool _executed;
-  bool _is_direct;
+class HandshakeOperation : public CHeapObj<mtThread> {
+  friend class HandshakeState;
+ protected:
+  HandshakeClosure*   _handshake_cl;
+  int32_t             _pending_threads;
+  JavaThread*         _target;
 public:
-  HandshakeOperation(HandshakeClosure* cl, bool is_direct = false) :
+  HandshakeOperation(HandshakeClosure* cl, JavaThread* target) :
     _handshake_cl(cl),
     _pending_threads(1),
-    _executed(false),
-    _is_direct(is_direct) {}
-
+    _target(target) {}
+  virtual ~HandshakeOperation() {}
   void do_handshake(JavaThread* thread);
   bool is_completed() {
     int32_t val = Atomic::load(&_pending_threads);
     assert(val >= 0, "_pending_threads=%d cannot be negative", val);
     return val == 0;
   }
-  void add_target_count(int count) { Atomic::add(&_pending_threads, count, memory_order_relaxed); }
-  bool executed() const { return _executed; }
-  const char* name() { return _handshake_cl->name(); }
+  void add_target_count(int count) { Atomic::add(&_pending_threads, count); }
+  const char* name()               { return _handshake_cl->name(); }
+  bool is_asynch()                 { return _handshake_cl->is_asynch(); }
+};
 
-  bool is_direct() { return _is_direct; }
+class AsyncHandshakeOperation : public HandshakeOperation {
+ private:
+  jlong _start_time_ns;
+ public:
+  AsyncHandshakeOperation(AsynchHandshakeClosure* cl, JavaThread* target, jlong start_ns)
+    : HandshakeOperation(cl, target), _start_time_ns(start_ns) {}
+  virtual ~AsyncHandshakeOperation() { delete _handshake_cl; };
+  jlong start_time()                 { return _start_time_ns; }
```

Comment:
Should this be 'const'? Ignore it if it would fan out too much.

Reply:
Fixed